File: src/secdashboards/detections/runner.py
"""Detection rule runner for executing rules against data sources."""

import importlib
import importlib.util
import os
import sys
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

import structlog
import yaml

from secdashboards.catalog.registry import DataCatalog
from secdashboards.connectors.base import DataConnector
from secdashboards.detections.rule import (
    DetectionMetadata,
    DetectionResult,
    DetectionRule,
    DualTargetDetectionRule,
    QueryTarget,
    Severity,
    SQLDetectionRule,
)

logger = structlog.get_logger()

# Environment variable to control Python rule loading
# Set SECDASH_ALLOW_PYTHON_RULES=1 to enable (NOT recommended in production)
ALLOW_PYTHON_RULES = os.environ.get("SECDASH_ALLOW_PYTHON_RULES", "").lower() in (
    "1",
    "true",
    "yes",
)


class DetectionRunner:
    """Runs detection rules against data sources.

    Security Notes:
    - By default, only YAML rules are loaded from filesystem
    - Python rule loading is disabled unless SECDASH_ALLOW_PYTHON_RULES=1
    - For production, use S3RuleStore for secure rule management
    """

    def __init__(
        self,
        catalog: DataCatalog,
        allow_python_rules: bool | None = None,
    ) -> None:
        """Initialize the detection runner.

        Args:
            catalog: Data catalog for source resolution
            allow_python_rules: Override env var for Python rules (None = use env)
        """
        self.catalog = catalog
        self._rules: dict[str, DetectionRule] = {}
        self._allow_python_rules = (
            allow_python_rules if allow_python_rules is not None else ALLOW_PYTHON_RULES
        )

        if self._allow_python_rules:
            logger.warning(
                "python_rules_enabled",
                message="Python rule loading is enabled. This allows arbitrary code execution.",
            )

    def register_rule(self, rule: DetectionRule) -> None:
        """Register a detection rule."""
        self._rules[rule.id] = rule
        logger.info("Registered detection rule", rule_id=rule.id, rule_name=rule.name)

    def get_rule(self, rule_id: str) -> DetectionRule | None:
        """Get a rule by ID."""
        return self._rules.get(rule_id)

    def list_rules(self, enabled_only: bool = True) -> list[DetectionRule]:
        """List all registered rules."""
        rules = list(self._rules.values())
        if enabled_only:
            rules = [r for r in rules if r.metadata.enabled]
        return rules

    def run_rule(
        self,
        rule_id: str,
        connector: DataConnector,
        start: datetime | None = None,
        end: datetime | None = None,
        lookback_minutes: int = 15,
    ) -> DetectionResult:
        """Run a single detection rule."""
        rule = self._rules.get(rule_id)
        if not rule:
            return DetectionResult(
                rule_id=rule_id,
                rule_name="Unknown",
                triggered=False,
                error=f"Rule not found: {rule_id}",
            )

        # Set time window
        end = end or datetime.now(UTC)
        start = start or (end - timedelta(minutes=lookback_minutes))

        try:
            # Get and execute query
            query = rule.get_query(start, end)
            logger.debug("Executing detection query", rule_id=rule_id, query=query[:200])

            df = connector.query(query)

            # Evaluate results
            result = rule.evaluate(df)
            logger.info(
                "Detection completed",
                rule_id=rule_id,
                triggered=result.triggered,
                match_count=result.match_count,
            )
            return result

        except Exception as e:
            logger.exception("Detection rule failed", rule_id=rule_id)
            return DetectionResult(
                rule_id=rule_id,
                rule_name=rule.name,
                triggered=False,
                error=str(e),
            )

    def run_all(
        self,
        connector: DataConnector,
        start: datetime | None = None,
        end: datetime | None = None,
        lookback_minutes: int = 15,
    ) -> list[DetectionResult]:
        """Run all enabled detection rules."""
        results = []
        for rule_id in self._rules:
            rule = self._rules[rule_id]
            if rule.metadata.enabled:
                result = self.run_rule(rule_id, connector, start, end, lookback_minutes)
                results.append(result)
        return results

    def load_rules_from_directory(self, rules_dir: Path) -> int:
        """Load detection rules from YAML files in a directory.

        Security Note: Python rules are only loaded if explicitly enabled
        via SECDASH_ALLOW_PYTHON_RULES=1 or allow_python_rules=True.
        """
        loaded = 0

        # Always load YAML rules (safe)
        for yaml_file in rules_dir.glob("*.yaml"):
            try:
                rules = self._load_yaml_rules(yaml_file)
                for rule in rules:
                    self.register_rule(rule)
                    loaded += 1
            except Exception as e:
                logger.exception(
                    "Failed to load rules from file", file=str(yaml_file), error=str(e)
                )

        # Only load Python rules if explicitly enabled
        if self._allow_python_rules:
            for py_file in rules_dir.glob("*.py"):
                if py_file.name.startswith("_"):
                    continue
                try:
                    rules = self._load_python_rules(py_file)
                    for rule in rules:
                        self.register_rule(rule)
                        loaded += 1
                except Exception as e:
                    logger.exception(
                        "Failed to load rules from file", file=str(py_file), error=str(e)
                    )
        else:
            # Log skipped Python files for visibility
            py_files = list(rules_dir.glob("*.py"))
            if py_files:
                logger.info(
                    "skipped_python_rules",
                    count=len(py_files),
                    message="Python rules disabled. Set SECDASH_ALLOW_PYTHON_RULES=1 to enable.",
                )

        return loaded

    def _load_yaml_rules(self, file_path: Path) -> list[DetectionRule]:
        """Load detection rules from a YAML file.

        Supports:
        - SQLDetectionRule: Traditional SQL/Athena queries
        - DualTargetDetectionRule: Rules with both CloudWatch and Athena queries

        Format detection:
        - If 'queries' dict is present → DualTargetDetectionRule
        - If 'log_type: cloudwatch_logs' → DualTargetDetectionRule (CW only)
        - Otherwise → SQLDetectionRule (Athena)
        """
        with file_path.open() as f:
            data = yaml.safe_load(f)

        rules: list[DetectionRule] = []
        rule_defs = data if isinstance(data, list) else [data]

        for rule_def in rule_defs:
            # Determine rule type based on structure
            has_queries_dict = "queries" in rule_def
            log_type = rule_def.get("log_type", "")
            is_cloudwatch = log_type == "cloudwatch_logs"

            if has_queries_dict or is_cloudwatch:
                # Use DualTargetDetectionRule for CloudWatch-based rules
                rule = self._parse_dual_target_rule(rule_def)
            else:
                # Use traditional SQLDetectionRule for Athena-only rules
                metadata = DetectionMetadata(
                    id=rule_def["id"],
                    name=rule_def["name"],
                    description=rule_def.get("description", ""),
                    author=rule_def.get("author", ""),
                    severity=Severity(rule_def.get("severity", "medium")),
                    tags=rule_def.get("tags", []),
                    mitre_attack=rule_def.get("mitre_attack", []),
                    data_sources=rule_def.get("data_sources", []),
                    schedule=rule_def.get("schedule", "rate(5 minutes)"),
                    enabled=rule_def.get("enabled", True),
                )

                rule = SQLDetectionRule(
                    metadata=metadata,
                    query_template=rule_def["query"],
                    threshold=rule_def.get("threshold", 1),
                    group_by_fields=rule_def.get("group_by", []),
                )

            rules.append(rule)

        return rules

    def _load_python_rules(self, file_path: Path) -> list[DetectionRule]:
        """Load custom detection rules from a Python file."""
        # Add parent directory to path for imports
        sys.path.insert(0, str(file_path.parent))

        try:
            spec = importlib.util.spec_from_file_location(file_path.stem, file_path)
            if not spec or not spec.loader:
                return []

            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)

            # Find all DetectionRule subclasses in the module
            rules = []
            for name in dir(module):
                obj = getattr(module, name)
                if (
                    isinstance(obj, type)
                    and issubclass(obj, DetectionRule)
                    and obj is not DetectionRule
                    and obj is not SQLDetectionRule
                ):
                    # Check if there's a corresponding instance or factory
                    instance_name = name.lower().replace("rule", "") + "_rule"
                    if hasattr(module, instance_name):
                        rules.append(getattr(module, instance_name))

            # Also check for rule instances directly
            for name in dir(module):
                obj = getattr(module, name)
                if isinstance(obj, DetectionRule):
                    rules.append(obj)

            return rules

        finally:
            sys.path.pop(0)

    def export_rules_to_dict(self) -> list[dict[str, Any]]:
        """Export all rules to a list of dictionaries."""
        return [rule.to_dict() for rule in self._rules.values()]

    def run_rule_with_target(
        self,
        rule_id: str,
        connector: DataConnector,
        target: QueryTarget,
        start: datetime | None = None,
        end: datetime | None = None,
        lookback_minutes: int = 15,
    ) -> DetectionResult:
        """Run a detection rule against a specific target.

        This is useful for DualTargetDetectionRule to choose between
        CloudWatch (hot tier) and Athena/Security Lake (cold tier).

        Args:
            rule_id: Rule identifier
            connector: Data connector (CloudWatchLogsConnector or AthenaConnector)
            target: Query target to use
            start: Query start time
            end: Query end time
            lookback_minutes: Lookback window if start not specified
        """
        rule = self._rules.get(rule_id)
        if not rule:
            return DetectionResult(
                rule_id=rule_id,
                rule_name="Unknown",
                triggered=False,
                error=f"Rule not found: {rule_id}",
            )

        end = end or datetime.now(UTC)
        start = start or (end - timedelta(minutes=lookback_minutes))

        try:
            # Get query for specific target
            if isinstance(rule, DualTargetDetectionRule):
                query = rule.get_query_for_target(target, start, end)
            else:
                # Fall back to default query for non-dual rules
                query = rule.get_query(start, end)

            logger.debug(
                "Executing detection query",
                rule_id=rule_id,
                target=target.value,
                query=query[:200],
            )

            # Execute query
            if target == QueryTarget.CLOUDWATCH:
                # CloudWatch connector uses query_insights with time params
                from secdashboards.connectors.cloudwatch_logs import CloudWatchLogsConnector

                if isinstance(connector, CloudWatchLogsConnector):
                    df = connector.query_insights(query, start=start, end=end)
                else:
                    df = connector.query(query)
            else:
                df = connector.query(query)

            # Evaluate results
            result = rule.evaluate(df)
            result_dict = result.model_dump() if hasattr(result, "model_dump") else {}
            result_dict["target"] = target.value

            logger.info(
                "Detection completed",
                rule_id=rule_id,
                target=target.value,
                triggered=result.triggered,
                match_count=result.match_count,
            )
            return result

        except Exception as e:
            logger.exception("Detection rule failed", rule_id=rule_id, target=target.value)
            return DetectionResult(
                rule_id=rule_id,
                rule_name=rule.name,
                triggered=False,
                error=str(e),
            )

    def run_dual_target(
        self,
        rule_id: str,
        cw_connector: DataConnector | None,
        athena_connector: DataConnector | None,
        hot_tier_hours: int = 168,  # 7 days
        lookback_minutes: int = 15,
    ) -> list[DetectionResult]:
        """Run a dual-target rule against both hot and cold tiers.

        This executes the rule against:
        1. CloudWatch Logs (hot tier) for recent data
        2. Athena/Security Lake (cold tier) for historical data

        Args:
            rule_id: Rule identifier
            cw_connector: CloudWatch Logs connector for hot tier
            athena_connector: Athena/Security Lake connector for cold tier
            hot_tier_hours: Hours of data in hot tier (default 7 days)
            lookback_minutes: Lookback window for detection

        Returns:
            List of DetectionResults (one per target executed)
        """
        rule = self._rules.get(rule_id)
        if not rule:
            return [
                DetectionResult(
                    rule_id=rule_id,
                    rule_name="Unknown",
                    triggered=False,
                    error=f"Rule not found: {rule_id}",
                )
            ]

        results = []
        end = datetime.now(UTC)
        start = end - timedelta(minutes=lookback_minutes)

        # Run against CloudWatch if available and rule supports it
        if (
            cw_connector
            and isinstance(rule, DualTargetDetectionRule)
            and rule.has_target(QueryTarget.CLOUDWATCH)
        ):
            result = self.run_rule_with_target(
                rule_id, cw_connector, QueryTarget.CLOUDWATCH, start, end
            )
            results.append(result)

        # Run against Athena/Security Lake if available
        if athena_connector:
            if isinstance(rule, DualTargetDetectionRule):
                if rule.has_target(QueryTarget.ATHENA):
                    result = self.run_rule_with_target(
                        rule_id, athena_connector, QueryTarget.ATHENA, start, end
                    )
                    results.append(result)
            else:
                # Non-dual rules default to Athena
                result = self.run_rule(rule_id, athena_connector, start, end)
                results.append(result)

        return results

    def _parse_dual_target_rule(self, rule_def: dict[str, Any]) -> DualTargetDetectionRule:
        """Parse a dual-target rule from YAML definition."""
        metadata = DetectionMetadata(
            id=rule_def["id"],
            name=rule_def["name"],
            description=rule_def.get("description", ""),
            author=rule_def.get("author", ""),
            severity=Severity(rule_def.get("severity", "medium")),
            tags=rule_def.get("tags", []),
            mitre_attack=rule_def.get("mitre_attack", []),
            data_sources=rule_def.get("data_sources", []),
            schedule=rule_def.get("schedule", "rate(5 minutes)"),
            enabled=rule_def.get("enabled", True),
        )

        # Handle both 'queries' dict and legacy 'query' string
        queries = rule_def.get("queries", {})
        if not queries and "query" in rule_def:
            # Legacy single-query format
            log_type = rule_def.get("log_type", "athena")
            if log_type == "cloudwatch_logs":
                queries = {"cloudwatch": rule_def["query"]}
            else:
                queries = {"athena": rule_def["query"]}

        return DualTargetDetectionRule(
            metadata=metadata,
            queries=queries,
            threshold=rule_def.get("threshold", 1),
            group_by_fields=rule_def.get("group_by", []),
        )

File: iris-rs/crates/iris-core/src/detections/runner.rs
use std::collections::HashMap;
use std::path::Path;

use chrono::{DateTime, Duration, Utc};
use tracing::{debug, info, warn};

use super::rule::{
    DetectionMetadata, DetectionResult, DetectionRule, DualTargetDetectionRule, SQLDetectionRule,
    Severity,
};
use crate::connectors::base::DataConnector;

/// Runs detection rules against data source connectors.
///
/// Only YAML rules are loaded from the filesystem — no arbitrary code execution.
pub struct DetectionRunner {
    rules: HashMap<String, Box<dyn DetectionRule>>,
}

impl DetectionRunner {
    /// Create an empty runner.
    #[must_use]
    pub fn new() -> Self {
        Self {
            rules: HashMap::new(),
        }
    }

    /// Register a detection rule.
    pub fn register_rule(&mut self, rule: Box<dyn DetectionRule>) {
        info!(
            rule_id = rule.id(),
            rule_name = rule.name(),
            "Registered detection rule"
        );
        self.rules.insert(rule.id().to_string(), rule);
    }

    /// Get a rule by ID.
    #[must_use]
    pub fn get_rule(&self, rule_id: &str) -> Option<&dyn DetectionRule> {
        self.rules.get(rule_id).map(AsRef::as_ref)
    }

    /// List all registered rules, optionally filtering to enabled only.
    #[must_use]
    pub fn list_rules(&self, enabled_only: bool) -> Vec<&dyn DetectionRule> {
        self.rules
            .values()
            .filter(|r| !enabled_only || r.metadata().enabled)
            .map(AsRef::as_ref)
            .collect()
    }

    /// Run a single detection rule.
    pub fn run_rule(
        &self,
        rule_id: &str,
        connector: &dyn DataConnector,
        start: Option<DateTime<Utc>>,
        end: Option<DateTime<Utc>>,
        lookback_minutes: i64,
    ) -> DetectionResult {
        let Some(rule) = self.rules.get(rule_id) else {
            return DetectionResult::not_found(rule_id);
        };

        let end = end.unwrap_or_else(Utc::now);
        let start = start.unwrap_or_else(|| end - Duration::minutes(lookback_minutes));

        let query = rule.get_query(start, end);
        debug!(
            rule_id,
            query_preview = &query[..query.len().min(200)],
            "Executing detection query"
        );

        match connector.query(&query) {
            Ok(qr) => {
                let result = rule.evaluate(&qr);
                info!(
                    rule_id,
                    triggered = result.triggered,
                    match_count = result.match_count,
                    "Detection completed"
                );
                result
            }
            Err(e) => {
                warn!(rule_id, error = %e, "Detection rule failed");
                DetectionResult::error(rule_id, rule.name(), &e.to_string())
            }
        }
    }

    /// Run all enabled detection rules.
    pub fn run_all(
        &self,
        connector: &dyn DataConnector,
        start: Option<DateTime<Utc>>,
        end: Option<DateTime<Utc>>,
        lookback_minutes: i64,
    ) -> Vec<DetectionResult> {
        self.rules
            .iter()
            .filter(|(_, r)| r.metadata().enabled)
            .map(|(id, _)| self.run_rule(id, connector, start, end, lookback_minutes))
            .collect()
    }

    /// Load detection rules from YAML files in a directory.
    ///
    /// Returns the number of rules successfully loaded.
    pub fn load_rules_from_directory(&mut self, rules_dir: &Path) -> usize {
        let mut loaded = 0;

        let entries = match std::fs::read_dir(rules_dir) {
            Ok(entries) => entries,
            Err(e) => {
                warn!(dir = %rules_dir.display(), error = %e, "Failed to read rules directory");
                return 0;
            }
        };

        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) != Some("yaml") {
                continue;
            }

            match load_yaml_rules(&path) {
                Ok(rules) => {
                    for rule in rules {
                        self.register_rule(rule);
                        loaded += 1;
                    }
                }
                Err(e) => {
                    warn!(file = %path.display(), error = %e, "Failed to load rules from file");
                }
            }
        }

        loaded
    }

    /// Export all rules to a list of JSON maps.
    #[must_use]
    pub fn export_rules_to_dict(&self) -> Vec<serde_json::Map<String, serde_json::Value>> {
        self.rules.values().map(|r| r.to_dict()).collect()
    }
}

impl Default for DetectionRunner {
    fn default() -> Self {
        Self::new()
    }
}

/// Load detection rules from a YAML file.
///
/// Supports:
/// - `SQLDetectionRule`: Traditional SQL/Athena queries (has `query` field)
/// - `DualTargetDetectionRule`: Rules with `queries` dict or `log_type: cloudwatch_logs`
fn load_yaml_rules(
    file_path: &Path,
) -> Result<Vec<Box<dyn DetectionRule>>, Box<dyn std::error::Error>> {
    let content = std::fs::read_to_string(file_path)?;
    let data: serde_yaml::Value = serde_yaml::from_str(&content)?;

    let rule_defs: Vec<serde_yaml::Value> = match data {
        serde_yaml::Value::Sequence(seq) => seq,
        other => vec![other],
    };

    let mut rules: Vec<Box<dyn DetectionRule>> = Vec::new();

    for rule_def in rule_defs {
        let map = rule_def
            .as_mapping()
            .ok_or("Rule definition must be a mapping")?;

        let has_queries = map.contains_key(yaml_key("queries"));
        let log_type = map
            .get(yaml_key("log_type"))
            .and_then(serde_yaml::Value::as_str)
            .unwrap_or("");

        let meta = parse_metadata(map)?;
        #[allow(clippy::cast_possible_truncation)]
        let threshold = map
            .get(yaml_key("threshold"))
            .and_then(serde_yaml::Value::as_u64)
            .unwrap_or(1) as usize;
        let group_by = map
            .get(yaml_key("group_by"))
            .and_then(serde_yaml::Value::as_sequence)
            .map(|seq| {
                seq.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default();

        if has_queries || log_type == "cloudwatch_logs" {
            let raw_queries = parse_queries(map, log_type);
            let rule = DualTargetDetectionRule::new(meta, raw_queries, threshold, group_by);
            rules.push(Box::new(rule));
        } else {
            let query = map
                .get(yaml_key("query"))
                .and_then(serde_yaml::Value::as_str)
                .unwrap_or("")
                .to_string();
            let rule = SQLDetectionRule {
                meta,
                query_template: query,
                threshold,
                group_by_fields: group_by,
            };
            rules.push(Box::new(rule));
        }
    }

    Ok(rules)
}

/// Create a YAML string key for map lookups.
fn yaml_key(key: &str) -> serde_yaml::Value {
    serde_yaml::Value::String(key.into())
}

fn parse_metadata(
    map: &serde_yaml::Mapping,
) -> Result<DetectionMetadata, Box<dyn std::error::Error>> {
    let get_str = |key: &str| -> String {
        map.get(yaml_key(key))
            .and_then(serde_yaml::Value::as_str)
            .unwrap_or("")
            .to_string()
    };
    let get_str_vec = |key: &str| -> Vec<String> {
        map.get(yaml_key(key))
            .and_then(serde_yaml::Value::as_sequence)
            .map(|seq| {
                seq.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default()
    };

    let id = get_str("id");
    if id.is_empty() {
        return Err("Rule must have an 'id' field".into());
    }
    let name = get_str("name");
    if name.is_empty() {
        return Err("Rule must have a 'name' field".into());
    }

    let severity_str = get_str("severity");
    let severity: Severity = if severity_str.is_empty() {
        Severity::Medium
    } else {
        serde_json::from_value(serde_json::Value::String(severity_str))?
    };

    let enabled = map
        .get(yaml_key("enabled"))
        .and_then(serde_yaml::Value::as_bool)
        .unwrap_or(true);

    Ok(DetectionMetadata {
        id,
        name,
        description: get_str("description"),
        author: get_str("author"),
        severity,
        tags: get_str_vec("tags"),
        mitre_attack: get_str_vec("mitre_attack"),
        data_sources: get_str_vec("data_sources"),
        schedule: {
            let s = get_str("schedule");
            if s.is_empty() {
                "rate(5 minutes)".into()
            } else {
                s
            }
        },
        enabled,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    })
}

fn parse_queries(map: &serde_yaml::Mapping, log_type: &str) -> HashMap<String, String> {
    // Try 'queries' dict first
    if let Some(queries_val) = map.get(yaml_key("queries"))
        && let Some(queries_map) = queries_val.as_mapping()
    {
        return queries_map
            .iter()
            .filter_map(|(k, v)| Some((k.as_str()?.to_string(), v.as_str()?.to_string())))
            .collect();
    }

    // Legacy: single 'query' field
    if let Some(query) = map
        .get(yaml_key("query"))
        .and_then(serde_yaml::Value::as_str)
    {
        let key = if log_type == "cloudwatch_logs" {
            "cloudwatch"
        } else {
            "athena"
        };
        let mut queries = HashMap::new();
        queries.insert(key.to_string(), query.to_string());
        return queries;
    }

    HashMap::new()
}

#[cfg(test)]
mod tests {
    use std::io::Write;

    use super::*;
    use crate::connectors::result::QueryResult;
    use crate::json_row;

    /// A mock connector that returns preset results.
    struct MockConnector {
        result: QueryResult,
    }

    impl DataConnector for MockConnector {
        fn query(
            &self,
            _sql: &str,
        ) -> Result<QueryResult, Box<dyn std::error::Error + Send + Sync>> {
            Ok(self.result.clone())
        }
        fn get_schema(
            &self,
        ) -> Result<HashMap<String, String>, Box<dyn std::error::Error + Send + Sync>> {
            Ok(HashMap::new())
        }
        fn check_health(
            &self,
        ) -> Result<
            crate::connectors::base::HealthCheckResult,
            Box<dyn std::error::Error + Send + Sync>,
        > {
            Ok(crate::connectors::base::HealthCheckResult::new(
                "mock", true,
            ))
        }
    }

    fn sample_sql_rule() -> SQLDetectionRule {
        SQLDetectionRule {
            meta: DetectionMetadata {
                id: "test-rule".into(),
                name: "Test Rule".into(),
                severity: Severity::High,
                enabled: true,
                ..serde_json::from_str::<DetectionMetadata>(
                    r#"{"id":"test-rule","name":"Test Rule","severity":"high"}"#,
                )
                .unwrap()
            },
            query_template: "SELECT * FROM t WHERE time >= TIMESTAMP '{start_time}'".into(),
            threshold: 1,
            group_by_fields: Vec::new(),
        }
    }

    #[test]
    fn register_and_get_rule() {
        let mut runner = DetectionRunner::new();
        runner.register_rule(Box::new(sample_sql_rule()));
        assert!(runner.get_rule("test-rule").is_some());
        assert!(runner.get_rule("missing").is_none());
    }

    #[test]
    fn list_rules_enabled_only() {
        let mut runner = DetectionRunner::new();
        let mut rule = sample_sql_rule();
        runner.register_rule(Box::new(rule.clone()));

        rule.meta.id = "disabled-rule".into();
        rule.meta.name = "Disabled Rule".into();
        rule.meta.enabled = false;
        runner.register_rule(Box::new(rule));

        assert_eq!(runner.list_rules(true).len(), 1);
        assert_eq!(runner.list_rules(false).len(), 2);
    }

    #[test]
    fn run_rule_not_found() {
        let runner = DetectionRunner::new();
        let connector = MockConnector {
            result: QueryResult::empty(),
        };
        let result = runner.run_rule("missing", &connector, None, None, 15);
        assert!(result.error.is_some());
    }

    #[test]
    fn run_rule_triggers() {
        let mut runner = DetectionRunner::new();
        runner.register_rule(Box::new(sample_sql_rule()));

        let connector = MockConnector {
            result: QueryResult::from_maps(vec![
                json_row!("user" => "alice"),
                json_row!("user" => "bob"),
            ]),
        };
        let result = runner.run_rule("test-rule", &connector, None, None, 15);
        assert!(result.triggered);
        assert_eq!(result.match_count, 2);
    }

    #[test]
    fn run_all_runs_enabled_rules() {
        let mut runner = DetectionRunner::new();
        runner.register_rule(Box::new(sample_sql_rule()));

        let connector = MockConnector {
            result: QueryResult::from_maps(vec![json_row!("x" => 1)]),
        };
        let results = runner.run_all(&connector, None, None, 15);
        assert_eq!(results.len(), 1);
        assert!(results[0].triggered);
    }

    #[test]
    fn export_rules_to_dict() {
        let mut runner = DetectionRunner::new();
        runner.register_rule(Box::new(sample_sql_rule()));
        let dicts = runner.export_rules_to_dict();
        assert_eq!(dicts.len(), 1);
        assert_eq!(dicts[0]["id"], "test-rule");
    }

    #[test]
    fn load_yaml_rules_from_file() {
        let dir = tempfile::tempdir().unwrap();
        let file_path = dir.path().join("rules.yaml");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(
            f,
            r#"- id: yaml-rule-1
  name: YAML Rule 1
  severity: high
  threshold: 1
  query: "SELECT * FROM t WHERE time >= TIMESTAMP '{{start_time}}'"
- id: yaml-rule-2
  name: YAML Rule 2
  severity: low
  threshold: 5
  query: "SELECT count(*) FROM t"
"#
        )
        .unwrap();

        let mut runner = DetectionRunner::new();
        let loaded = runner.load_rules_from_directory(dir.path());
        assert_eq!(loaded, 2);
        assert!(runner.get_rule("yaml-rule-1").is_some());
        assert!(runner.get_rule("yaml-rule-2").is_some());
    }

    #[test]
    fn load_yaml_dual_target_rule() {
        let dir = tempfile::tempdir().unwrap();
        let file_path = dir.path().join("dual.yaml");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(
            f,
            r#"- id: dual-rule
  name: Dual Target Rule
  severity: medium
  queries:
    cloudwatch: "fields @timestamp | filter @message like /ERROR/"
    athena: "SELECT * FROM t WHERE severity >= 4"
"#
        )
        .unwrap();

        let mut runner = DetectionRunner::new();
        let loaded = runner.load_rules_from_directory(dir.path());
        assert_eq!(loaded, 1);

        let rule = runner.get_rule("dual-rule").unwrap();
        assert_eq!(rule.name(), "Dual Target Rule");
    }
}
